### PR TITLE
build: Finish up out-of-tree changes

### PR DESCRIPTION
**Pull request by theuni**
This fixes the remaining out-of-tree issues:
- The entire leveldb dir was copied out
- Python caches were left hanging around
- A qt file was left dangling in the build-dir
- The rpc tests couldn't be run
- OSX's `make deploy` was broken

`make distcheck` now passes, and Travis has been switched to an out-of-tree build for dogfooding purposes. Fingers-crossed that it works.

### CODE REVIEW

File: qa/pull-tester/rpc-tests.py [hunk 1]
```diff
@@ -37,7 +38,7 @@
     # terminal via ANSI escape sequences:
     BOLD = ('\033[0m', '\033[1m')
 
-RPC_TESTS_DIR = BUILDDIR + '/qa/rpc-tests/'
+RPC_TESTS_DIR = SRCDIR + '/qa/rpc-tests/'
```
Comment (MarcoFalke) [inline qa/pull-tester/rpc-tests.py]: Note to myself: Does the cache (from `create_cache.py` et al.) end up in SRCDIR now?

**Comment by MarcoFalke** [comment]
Concept ACK

**Comment by theuni** [comment]
Generally speaking, if it's used as-is from git, it's srcdir. If it's generated or pre-processed, it's builddir.

Put another way: after checkout, think of srcdir as read-only. If anything is modified, it must happen in builddir.

**Comment by laanwj** [comment]
utACK 142ffc7

File: qa/pull-tester/rpc-tests.py (hunk 1 above)
Comment (laanwj) [inline qa/pull-tester/rpc-tests.py]: Wasn't it always (and still is) the case that the cache ends up in the current directory when the tests are launched?

Comment (MarcoFalke) [inline qa/pull-tester/rpc-tests.py]: Correct.

Though, this should be changed. Having different caches lying around in folders that happen to be your working dir right then is misleading at best.

Anyway, this is unrelated to this pull.

Comment (theuni) [inline qa/pull-tester/rpc-tests.py]: I'm having a hard time answering that question because of the symlink interaction with Python. builddir/rpc-tests.py is now a symlink to sourcedir/rpc-tests.py because it needs to be writable. But that makes the definition of "current dir" hazy.

@MarcoFalke I suppose it would be best to make the location explicit.

Comment (MarcoFalke) [inline qa/pull-tester/rpc-tests.py]: Sure, this is on my list of cleanups. If no one beats me to it, you can expect it some time after feature freeze. ;)

**Comment by MarcoFalke** [comment]
`make check` fails with:

```py
Running test/bitcoin-util-test.py...
Traceback (most recent call last):
  File "../../src/test/bitcoin-util-test.py", line 8, in <module>
    import buildenv
ImportError: bad magic number in 'buildenv': b'\x03\xf3\r\n'

**Comment by theuni** [comment]
@MarcoFalke Yes, I just saw that as well while trying to answer the question above. I'm not sure why Travis passes.

I'll fix that one up.

**Comment by theuni** [comment]
Pushed a quick fix for @MarcoFalke's problem. I'm not sure if we really need that or not, but it's probably better than each dev running into the problem and googling for themselves.

The issue is this: previous invocations of python2 in an in-tree build would leave a buildenv.pyc hanging around in srcdir. We prepend python's search path with builddir, but for some reason it ends up finding the srcdir/buildenv.pyc (and missing .py) before the builddir/buildenv.py. python3 refuses to load the old python2 .pyc, and errors because it can't find a .py at the same path for recompilation.

With python3, pyc's are organized into a \__pycache__ dir, and tagged like: \__pycache__/buildenv.cpython-35.pyc. So it should not be possible to hit this case anymore.

**Comment by MarcoFalke** [comment]
Tested ACK 340012d

File: qa/pull-tester/rpc-tests.py [hunk 2]
```diff
@@ -29,6 +29,7 @@
 import tempfile
 import re
 
+sys.path.append("qa/pull-tester/")
```
Comment (MarcoFalke) [inline qa/pull-tester/rpc-tests.py]: Is it safe to assume everyone runs the tests from the root of buildir?

Comment (luke-jr) [inline qa/pull-tester/rpc-tests.py]: Probably not, but FWIW at least I do...

Comment (btcdrak) [inline qa/pull-tester/rpc-tests.py]: I do too.

Comment (sipa) [inline qa/pull-tester/rpc-tests.py]: Heh, I run them from qa/pull-tester. I had no idea you could run them from the root...

Comment (MarcoFalke) [inline qa/pull-tester/rpc-tests.py]: Not sure how to fix this...

What about byte-compiling the rpc-tests.py instead of linking?

**Comment by luke-jr** [comment]
Extensively-tested, with the following results:

I built, `make check`'d, and confirmed the reported version numbers for:
1. Before this PR (plus #7522), built in a clean git clone: v0.12.99.0-fd89d8e
2. Before this PR (plus #7522), built in a clean dist dir: v0.12.99.0-gfd89d8e
3. Before this PR (plus #7522), built in a clean dist dir under a foreign git checkout: v0.12.99.0-6862d12
4. With this PR (plus #7522), built in a clean git clone: v0.12.99.0-08cb06e
5. With this PR (plus #7522), built in a subdirectory of a clean git clone owned by another user: v0.12.99.0-08cb06e-dirty
6. With this PR (plus #7522), built in a parallel directory of a clean git clone owned by another user, under a foreign git checkout: v0.12.99.0-08cb06e-dirty
7. With this PR (plus #7522), built in a clean dist dir under a foreign git checkout: v0.12.99.0-6862d12
8. With this PR (plus #7522), built in a clean dist dir: v0.12.99.0-g08cb06e
9. With this PR (plus #7522), built in a subdirectory of a clean dist dir owned by another user: v0.12.99.0-g08cb06e
10. With this PR (plus #7522), built in a parallel directory of a clean dist dir owned by another user: v0.12.99.0-g08cb06e

Note that:
* 08cb06e = With this PR (plus #7522)
* fd89d8e = Before this PR (plus #7522)
* 6862d12 = Commit in empty foreign git repo

Observations:
1. No change to previously-supported build scenarios.
2. Dist dir builds within a foreign git repository incorrectly use the foreign git repo's commit hash in the version. This is already a problem (even with #7522 apparently!), so not a blocker for this PR.
3. Dist dir builds prefix the commit hash with 'g', while git builds do not. This is already a problem, so also not a regression in this PR.
4. Clean git builds using a different build dir incorrectly tag the version as "-dirty". **This should probably be fixed in this PR.**

**Comment by laanwj** [comment]
BTW seems `make translate` also needs changes to work out-of-tree. Not necessarily in this pull.

**Comment by laanwj** [comment]
Looks like this is enough:
```patch
diff --git a/src/Makefile.qt.include b/src/Makefile.qt.include
index 3b39919..651cccb 100644
--- a/src/Makefile.qt.include
+++ b/src/Makefile.qt.include
@@ -390,11 +390,11 @@ QT_QM=$(QT_TS:.ts=.qm)
 
 SECONDARY: $(QT_QM)
 
-qt/bitcoinstrings.cpp: $(libbitcoin_server_a_SOURCES) $(libbitcoin_wallet_a_SOURCES)
+$(srcdir)/qt/bitcoinstrings.cpp: $(libbitcoin_server_a_SOURCES) $(libbitcoin_wallet_a_SOURCES)
        @test -n $(XGETTEXT) || echo "xgettext is required for updating translations"
        $(AM_V_GEN) cd $(srcdir); XGETTEXT=$(XGETTEXT) PACKAGE_NAME="$(PACKAGE_NAME)" COPYRIGHT_HOLDERS="$(COPYRIGHT_HOLDERS)" COPYRIGHT_HOLDERS_SUBSTITUTION="$(COPYRIGHT_HOLDERS_SUBSTITUTION)
" $(PYTHON) ../share/qt/extract_strings_qt.py $^
 
-translate: qt/bitcoinstrings.cpp $(QT_FORMS_UI) $(QT_FORMS_UI) $(BITCOIN_QT_CPP) $(BITCOIN_QT_H) $(BITCOIN_MM)
+translate: $(srcdir)/qt/bitcoinstrings.cpp $(QT_FORMS_UI) $(QT_FORMS_UI) $(BITCOIN_QT_CPP) $(BITCOIN_QT_H) $(BITCOIN_MM)
        @test -n $(LUPDATE) || echo "lupdate is required for updating translations"
        $(AM_V_GEN) QT_SELECT=$(QT_SELECT) $(LUPDATE) $^ -locations relative -no-obsolete -ts qt/locale/bitcoin_en.ts
 
```

**Comment by theuni** [comment]
@luke-jr Whoa, that's great feedback! Thanks for the testing. I'm having a look at the "-dirty" issue.

@laanwj Thanks, I'll add that here.

**Comment by laanwj** [comment]
Let's try to get this merged soon.
I can PR the `make translate` change separately if that's mroe convenient.

**Comment by theuni** [comment]
@luke-jr I believe I've found the issue here. When you "chown -R root:root .", you're setting the git index to read-only. That makes "git diff" fail to update the index, which makes the diff-index fail.

For the sake of this PR, I'm ok with that not working correctly, since git is in a pretty bad state.

**Comment by laanwj** [comment]
> since git is in a pretty bad state

That sounds ominous. Time to switch to mercurial or bzr? :-)

**Comment by laanwj** [comment]
"works for me" ACK https://github.com/bitcoin/bitcoin/pull/8133/commits/d1a3d570e5f4e59683cde9ccc0ac10d012ef7070

**Comment by theuni** [comment]
Haha. That was a half-complete thought as usual.

Should read: "I'm ok with that not working correctly, since if you're hitting this case, your git repo is already unreliable".

File: qa/pull-tester/rpc-tests.py (hunk 2 above)
Comment (laanwj) [inline qa/pull-tester/rpc-tests.py]: > Heh, I run them from qa/pull-tester. I had no idea you could run them from the root...

Same here. But running from the root is ok with me.
